### PR TITLE
feat(api): delete workflow with dependencies

### PR DESCRIPTION
**Pull request by fsamin**

### CODE REVIEW

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#16560.0 ✘  
* Build
  * Build and Package All ✘  
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✔ 
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#16567.0 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✘  
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#16567.1 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✘  
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔

File: ui/src/app/views/workflow/show/admin/delete-modal/delete-modal.component.ts [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+    deleteWorkflow(b: boolean): void {
+        this.loading = true;
+        this.store.dispatch(
+            new DeleteWorkflow({
+                    projectKey: this.project.key,
+                    workflowName: this.workflow.name,
+                    withDependencies: b
+                })
+        ).pipe(
+            finalize(() => this.loading = false))
```
Comment (sguiheux) [inline ui/src/app/views/workflow/show/admin/delete-modal/delete-modal.component.ts:98]: missing this._cd.markForCheck();

File: ui/src/app/views/workflow/show/admin/delete-modal/delete-modal.component.ts [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+    closeCallback(): void {
+        this.open = false;
+    }
+
+    init(): void {
+        this._workflowService.getDeletedDependencies(this.workflow).pipe(
+            finalize(() => this.loading = false)
+        ).subscribe((x) => {
+            this.dependencies = x;
+            this._cd.markForCheck();
```
Comment (sguiheux) [inline ui/src/app/views/workflow/show/admin/delete-modal/delete-modal.component.ts:115]: In finalize

**Review by sguiheux — CHANGES_REQUESTED** [review]

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#16575.0 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✔ 
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔ 
* Package
  * Docker ✔ 
* Integration
  * Docker Compose Tests ✔ 
  * Standalone Tests ✘

File: engine/api/api_routes.go [hunk 3]
```diff
@@ -223,6 +223,7 @@ func (api *API) InitRouter() {
 	r.Handle("/project/{key}/type/{type}/access", ScopeNone(), r.GET(api.getProjectAccessHandler))
 	r.Handle("/project/{permProjectKey}/workflows", Scope(sdk.AuthConsumerScopeProject), r.POST(api.postWorkflowHandler), r.GET(api.getWorkflowsHandler))
 	r.Handle("/project/{key}/workflows/{permWorkflowName}", Scope(sdk.AuthConsumerScopeProject), r.GET(api.getWorkflowHandler), r.PUT(api.putWorkflowHandler), r.DELETE(api.deleteWorkflowHandler))
+	r.Handle("/project/{key}/workflows/{permWorkflowName}/delete/dependencies", Scope(sdk.AuthConsumerScopeProject), r.GET(api.getWorkflowDependencieswHandler))
```
Comment (richardlt) [inline engine/api/api_routes.go:226]:
```suggestion
	r.Handle("/project/{key}/workflows/{permWorkflowName}/delete/dependencies", Scope(sdk.AuthConsumerScopeProject), r.GET(api.getWorkflowDependenciesHandler))
```

File: engine/api/workflow.go [hunk 4]
```diff
@@ -930,3 +936,94 @@ func (api *API) getSearchWorkflowHandler() service.Handler {
 		return service.WriteJSON(w, ws, http.StatusOK)
 	}
 }
+
+func (api *API) getWorkflowDependencieswHandler() service.Handler {
```
Comment (richardlt) [inline engine/api/workflow.go:940]:
```suggestion
func (api *API) getWorkflowDependenciesHandler() service.Handler {
```

**Review by richardlt — CHANGES_REQUESTED** [review]

File: ui/src/app/views/workflow/show/admin/delete-modal/delete-modal.html [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+            </div>
+            <div class="row">
+                <div class="sixteen wide right aligned column">
+                    <app-delete-button title="Delete workflow and clean dependencies" [loading]="loading" (event)="deleteWorkflow(true)"></app-delete-button>
+                </div>
+            </div>
+            <div class="row">
+                <div class="sixteen wide column">
+                    <h3>Do you prefer to keep the dependencies ?</h3>
+                    <p>Only the workflow {{workflow.name}} will be deleted}</p>
```
Comment (richardlt) [inline ui/src/app/views/workflow/show/admin/delete-modal/delete-modal.html:39]:
```suggestion
                    <p>Only the workflow {{workflow.name}} will be deleted.</p>
```

File: ui/src/app/views/workflow/show/admin/delete-modal/delete-modal.html [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+                            <sm-list class="bulleted">
+                                <sm-item *ngFor="let pip of dependencies.deleted_dependencies.pipelines">pipeline <strong>{{pip.name}}</strong></sm-item>
+                                <sm-item *ngFor="let app of dependencies.deleted_dependencies.applications">application <strong>{{app.name}}</strong></sm-item>
+                                <sm-item *ngFor="let env of dependencies.deleted_dependencies.environments">environment <strong>{{env.name}}</strong></sm-item>
+                            </sm-list>
+                        </p>
+                    </ng-container>
+
+                    <ng-container *ngIf="dependencies?.unlinked_as_code_dependencies && !dependencies.unlinked_as_code_dependencies.isEmpty()">
+                        <p>The following objects will unlinked from repository "{{workflow.from_repository}}":
```
Comment (richardlt) [inline ui/src/app/views/workflow/show/admin/delete-modal/delete-modal.html:21]:
```suggestion
                        <p>The following objects will be unlinked from repository "{{workflow.from_repository}}":
```

File: ui/src/app/views/workflow/show/admin/delete-modal/delete-modal.html [hunk 7]
```diff
@@ -0,0 +1,49 @@
+<ng-template let-context let-modal="modal" #workflowDeleteModal>
+    <div class="header">
+        Deleting workflow {{workflow.name}}
+    </div>
+    <div class="content scrolling" [loading]="loading">
+        <div class="ui grid">
+            <div class="row">
+                <div class="sixteen wide column">
+                    <h3>Do you want to delete the workflow and clean its dependencies ?</h3>
```
Comment (richardlt) [inline ui/src/app/views/workflow/show/admin/delete-modal/delete-modal.html:9]: Can we disable the form "with dependencies" when no dependencies are found ?

**Review by richardlt — CHANGES_REQUESTED** [review]

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#16584.0 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✔ 
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔ 
* Package
  * Docker ✘

File: ui/src/app/views/workflow/show/admin/delete-modal/delete-modal.html (hunk 7 above)
Comment (fsamin) [reply ui/src/app/views/workflow/show/admin/delete-modal/delete-modal.html]: done

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#16585.0 ✘  
* Build
  * Build and Package All ✘  
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✔ 
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#16585.1 ✘  
* Build
  * Build and Package All ✘  
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✔ 
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#16585.2 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✔ 
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔ 
* Package
  * Docker ✘

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#16585.3 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✔ 
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔ 
* Package
  * Docker ✔ 
* Integration
  * Docker Compose Tests ✔ 
  * Standalone Tests ✘

**Review by richardlt — APPROVED** [review]

**Review by yesnault — APPROVED** [review]

**Review by sguiheux — APPROVED** [review]

**Pull request merged**